### PR TITLE
doc: enrich developer documentation

### LINKED ISSUE #1069: Write a developer documentation
In `docs/developer`, write documentation on:
- how to build an ISO (move the content from `BUILDING.md` here)
- how to use Vagrant, describe the setup, SSH keys, …

### PR DESCRIPTION
**Pull request by slaperche-scality**
**Component**:

documentation

**Context**: 

Enrich the developer documentation.

**Summary**:

Move `BUILDING.md` into the developer documentation.
Document the Vagrant setup.

**Acceptance criteria**: 

Doc builds and contains the new content

---

Closes: #1069

### CODE REVIEW

File: docs/developer/building/building.rst [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+Note that:
+- you can speed up the build by spawning more workers, e.g. ``./doit.sh -n 4``.
+- you can have a JSON output with ``./doit.sh --reporter json``
+
+When a task is prefixed by:
+- ``--``: the task is skipped because already up-to-date
+- ``.``: the task is executed
+- ``!!``: the task is ignored.
+
+## Main tasks
```
Comment (jbertran) [inline docs/developer/building/building.rst]: Stray markdown syntax

File: docs/developer/building/building.rst [hunk 2]
```diff
@@ -0,0 +1,33 @@
+How to build an ISO
+===================
+
+Our build system is based on `doit <http://pydoit.org/>`_.
+
+To build, simply type ``./doit.sh``.
+
+Note that:
+- you can speed up the build by spawning more workers, e.g. ``./doit.sh -n 4``.
```
Comment (jbertran) [inline docs/developer/building/building.rst:10]: Need 1 more line break so the list is rendered IIRC

File: docs/developer/building/building.rst [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+- you can have a JSON output with ``./doit.sh --reporter json``
+
+When a task is prefixed by:
+- ``--``: the task is skipped because already up-to-date
```
Comment (jbertran) [inline docs/developer/building/building.rst:15]: Ditto list

File: docs/developer/building/building.rst [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+- ``.``: the task is executed
+- ``!!``: the task is ignored.
+
+## Main tasks
+
+To get a list of the available targets, you can run ``./doit.sh list``.
+
+The most important ones are:
+- ``iso``:  build the MetalK8s ISO
```
Comment (jbertran) [inline docs/developer/building/building.rst:26]: Ditto list

File: docs/developer/building/building.rst [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+- ``lint``: run the linting tools on the codebase
+- ``populate_iso``: populate the ISO file tree
+- ``vagrant_up``: spawn a development environment using Vagrant
+
+By default, i.e. if you only type ``./doit.sh`` with no arguments, the ``iso``
+task is executed.
+
+You can also run a subset of the build only:
+- ``packaging``: download and build the software packages and repositories
```
Comment (jbertran) [inline docs/developer/building/building.rst:36]: Ditto list, you get the idea

File: docs/developer/building/features.rst [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    - _build_packages:calico-cni-plugin:pkg_mkdir
+
+   targets    :
+    - /home/foo/dev/metalk8s/_build/packages/calico-cni-plugin-3.5.1-1.el7.src.rpm
+
+Wildcard selection
+------------------
+
+You can use wildcard in task names, which allows you to either:
+- execute all the sub-tasks of a specific task:
```
Comment (jbertran) [inline docs/developer/building/features.rst:89]: And ditto on lists.

File: docs/developer/building/local_cluster.rst [hunk 7]
```diff
@@ -0,0 +1,24 @@
+Running a local cluster
+=======================
+
+You can spawn a local MetalK8s cluster by running ``./doit.sh vagrant_up``.
+
+This command will start a virtual machine (using VirtualBox) and:
+- mount the build tree
```
Comment (jbertran) [inline docs/developer/building/local_cluster.rst:8]: And ditto.

File: docs/developer/building/local_cluster.rst [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+- execute the bootstrap script to make this machine a bootstrap node
+
+After executing this command, you have a MetalK8s bootstrap node up and running
+and you can connect to it by using ``vagrant ssh bootstrap``.
+
+Note that you can extend your cluster by spawning extra nodes (up to 9 are
+already pre-defined in the provided ``Vagrantfile``) by running
+``vagrant up node1 --provision``.
+This will:
+- spawn a virtual machine for the node 1
```
Comment (jbertran) [inline docs/developer/building/local_cluster.rst:21]: And again.

File: docs/developer/building/requirements.rst [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  local and remote images
+- `hardlink <https://jak-linux.org/projects/hardlink/>`_: to de-duplicate images
+  layers
+- mkisofs: to create the MetalK8s ISO
+
+Optional
+--------
+
+- `git <https://git-scm.com/>`_: to add the Git reference in the build metadata
+- `Vagrant <https://www.vagrantup.com/>`_, 1.8 or higher: to spawn a local
```
Comment (jbertran) [inline docs/developer/building/requirements.rst:23]: We should convey that for now, Virtualbox is the only provider we support for the Vagrant side of things.

File: docs/developer/building/requirements.rst [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  cluster
+- `VirtualBox <https://www.virtualbox.org>`_: to spawn a local cluster
+- `tox <https://pypi.org/project/tox>`_: to run the linters
```
Comment (jbertran) [inline docs/developer/building/requirements.rst:26]: linters *and* end-to-end tests.

File: docs/developer/building/requirements.rst [hunk 11]
```diff
@@ -0,0 +1,32 @@
+Requirements
+============
+
+In order to build MetalK8s we rely and third-party tools, some of them are
+mandatory, others are optional.
+
+Mandatory
+---------
+
+- `Python <https://www.python.org/>`_ 3.6 or higher: our buildchain is
```
Comment (jbertran) [inline docs/developer/building/requirements.rst:10]: buildchain *and* BDD tests

**Review by jbertran — CHANGES_REQUESTED** [review]
Content's good, some RST issues :)

Thanks!

File: docs/developer/building/building.rst (hunk 1 above)
Comment (slaperche-scality) [reply docs/developer/building/building.rst]: Indeed

File: docs/developer/building/requirements.rst (hunk 10 above)
Comment (slaperche-scality) [reply docs/developer/building/requirements.rst:26]: But E2E tests are not run by the build chain so I only mentioned linters.

File: docs/developer/building/requirements.rst (hunk 11 above)
Comment (slaperche-scality) [reply docs/developer/building/requirements.rst:10]: This requirements is only about the build chain.

File: docs/developer/building/requirements.rst (hunk 9 above)
Comment (slaperche-scality) [reply docs/developer/building/requirements.rst:23]: Isn't that clear enough with the dependency on VirtualBox?
How can I formulate it in a clearer way?

**Comment by slaperche-scality** [comment]
@bert-e reset

File: docs/developer/building/requirements.rst (hunk 9 above)
Comment (jbertran) [reply docs/developer/building/requirements.rst:23]: I was hoping you'd have an idea :)

Maybe just tack on `Virtualbox is currently the only provider supported.` to the Vagrant item?

Comment (slaperche-scality) [reply docs/developer/building/requirements.rst:23]: Done

**Comment by slaperche-scality** [comment]
@bert-e reset

**Review by jbertran — APPROVED** [review]
LGTM!

**Comment by slaperche-scality** [comment]
@bert-e approve

**Pull request merged**